sys/cheri/cheri_cap_relocs.c

```diff
@@ -39,7 +40,65 @@ extern void init_cap_relocs(void *data_cap, void *code_cap);
 void
 init_cap_relocs(void *data_cap, void *code_cap)
 {
-	cheri_init_globals_3(data_cap, code_cap, data_cap);
+	const struct capreloc *start_relocs, *stop_relocs;
+
+	start_relocs = CHERI_RODATA_PTR(__start___cap_relocs);
+	stop_relocs = CHERI_RODATA_PTR(__stop___cap_relocs);
+
+	/*
+	 * TODO: Set code bounds if the ABI allows it (presence of
+	 * PT_CHERI_PCC)
+	 */
+	bool can_set_code_bounds = false;
+
+	for (const struct capreloc *reloc = start_relocs; reloc < stop_relocs;
+	     reloc++) {
+		const void **dest = __builtin_cheri_address_set(data_cap,
+		    reloc->capability_location);
+		void *src;
+		bool can_set_bounds = true;
+
+		if (reloc->object == 0) {
+			*dest = 0;
+			continue;
+		}
+
+		if ((reloc->permissions & function_reloc_flag) ==
+		    function_reloc_flag) {
+			/* code pointer */
+			src = __builtin_cheri_address_set(code_cap,
+			    reloc->object);
+			src = __builtin_cheri_perms_and(src,
+			    CHERI_PERMS_KERNEL_CODE);
+#ifdef CHERI_FLAGS_CAP_MODE
+			src = __builtin_cheri_flags_set(src,
+			    CHERI_FLAGS_CAP_MODE);
```

Comment:
This shouldn't be done for hybrid, but also the expectation is that code_cap already has the right flags (cheri_init_globals_3 doesn't set the flags).

Reply:
Note that this code is only used from purecap and not from hybrid (I guess we don't have any caprelocs in hybrid?)